basics.sh

```diff
@@ -47,51 +61,86 @@
+    # Here we are adding environment variables to the file for pyenv, one of the
+    # python virtual environment systems.
+
+    # For each line to be inserted into the file, we first check if the line
+    # already exists in the file using grep, and only if it's not present is it
+    # added in. This prevents these env variables being added twice if this
+    # script is ever re-run.
     touch ~/.zshrc
     if ! grep -Fxq 'export PYENV_ROOT="$HOME/.pyenv"' ~/.zshrc
     then
-        echo 'export PYENV_ROOT="$HOME/.pyenv"' >> ~/.zshrc
+        printf 'export PYENV_ROOT="$HOME/.pyenv"\n' >> ~/.zshrc
     fi
     if ! grep -Fxq 'export PATH="$PYENV_ROOT:$PATH"' ~/.zshrc
     then
-        echo 'export PATH="$PYENV_ROOT:$PATH"' >> ~/.zshrc
+        printf 'export PATH="$PYENV_ROOT:$PATH"\n' >> ~/.zshrc
     fi
+
+    # Insert the pyenv init command into zshrc if it does not already exist. The
+    # inserted code first checks if pyenv is installed and if present will
+    # configure pyenv as a shell function and enable shims and shell
+    # autocompletion.
     if ! grep -Fxq 'if command -v pyenv 1>/dev/null 2>&1; then' ~/.zshrc
     then
-        echo 'if command -v pyenv 1>/dev/null 2>&1; then\n  eval "$(pyenv init -)"\nfi' >> ~/.zshrc
+        printf 'if command -v pyenv 1>/dev/null 2>&1; then\n  eval "$(pyenv init -)"\nfi\n' >> ~/.zshrc
     fi
+
+    # Re-initialise the shell environment, loading the above shell variables
+    # and running the pyenv init command
     source ~/.zshrc
-    echo "\n-> installing software from homebrew casks:"
-    echo "this is a minimal list to get you started"
+
+    printf "\n-> installing software from homebrew casks:\n"
+    printf "this is a minimal list to get you started\n"
     {
         brew install --cask docker
-        brew install --cask dashlane
         brew install --cask dbeaver-community
         brew install --cask sublime-text
         brew install --cask the-unarchiver
         brew install --cask drawio
         pipx ensurepath && source ~/.zshrc
-    } && echo "\n-> installation from homebrew casks successfull <-\n"
-    echo "Adding compiler flags for bzip2"
-    echo "export LDFLAGS=-L/usr/local/opt/bzip2/lib" >> ~/.zshrc
-    echo "export CPPFLAGS=-I/usr/local/opt/bzip2/include" >> ~/.zshrc
-    echo "\n-> installing from pipx:"
-    echo "   pipenv pipenv-pipes pre-commit"
+    } && printf "\n-> installation from homebrew casks successfully <-\n\n"
+
+    # Insert compiler flags to the zshrc file only if they're not already
+    # present
+    printf "Adding compiler flags for bzip2\n"
+    if ! grep -Fxq 'export LDFLAGS=-L/usr/local/opt/bzip2/lib' ~/.zshrc
+    then
+        printf "export LDFLAGS=-L/usr/local/opt/bzip2/lib\n" >> ~/.zshrc
+    fi
+    if ! grep -Fxq 'export CPPFLAGS=-I/usr/local/opt/bzip2/include' ~/.zshrc
+        then
+    printf "export CPPFLAGS=-I/usr/local/opt/bzip2/include\n" >> ~/.zshrc
+    fi
+    printf "\n-> installing from pipx:\n"
```

Comment:
since you've started doing this elsewhere, I think it would be worth adding an empty line here - I won't leave this comment again but do please have a think about where else in the script it would make sense to do this.